feat: Add TypeScript Bindings and Improve Documentation

cuervolu opened the pull request:
## Changes in this PR

This pull request introduces several significant improvements to the OAuth plugin:

1. **TypeScript Bindings**: 
   - Created TypeScript bindings for the OAuth plugin, enabling easier integration in TypeScript-based Tauri applications.
   - Implemented functions like `start()`, `cancel()`, `onUrl()`, and `onInvalidUrl()` for TypeScript users.

2. **Updated Vanilla Example**:
   - Added a new example using the Tauri V2 vanilla template.
   - Demonstrates usage of the OAuth plugin in both Rust and TypeScript contexts.
   - Includes UI elements to start OAuth servers and display results.

3. **Improved README**:
   - Enhanced project description and rationale.
   - Added clear installation instructions for both Rust and npm/yarn.
   - Provided usage examples in Rust and TypeScript.
   - Included a configuration section with code samples.
   - Added important security considerations.

## Notes for Reviewers
- Please pay special attention to the TypeScript bindings to ensure they correctly mirror the Rust functionality.
- The vanilla example might benefit from additional error handling or edge case considerations.
- We need to verify that all code examples in the README work as expected. Please test these examples to ensure they are accurate and functional.
- An important next step is to publish the npm package for the TypeScript bindings. As i'm not familiar with this process, guidance or assistance with publishing to npm would be greatly appreciated.
- Any suggestions for further improving the documentation or examples are welcome!

cuervolu committed:
chore: Update tauri dependency to version 2.0.0-rc.0

cuervolu committed:
feat(ts): initialize TypeScript bindings for tauri-plugin-oauth

- Add package.json with TypeScript and Rollup configuration
- Set up project structure for TypeScript bindings
- Include both original author and TypeScript contributor
- Align npm package name with Tauri plugin naming conventions

cuervolu committed:
feat(bindings): add TypeScript bindings for OAuth plugin

- Create TypeScript interface for OauthConfig
- Implement start, cancel, onUrl, and onInvalidUrl functions
- Add OauthPlugin interface and oauth object
- Include JSDoc comments with examples for each function
- Ensure compatibility with Tauri's invoke and event system

cuervolu committed:
feat(example): add OAuth plugin demo using Tauri V2 vanilla template

This commit introduces a new example demonstrating the usage of the OAuth plugin
with the Tauri V2 vanilla template. The example includes:

- HTML interface with buttons to start OAuth servers using Rust and TypeScript
- TypeScript code to manage OAuth server states and handle user interactions
- Rust backend modifications to support starting and stopping OAuth servers

cuervolu committed:
docs(readme): enhance project documentation and examples

This commit significantly improves the README.md file by:

- Adding a clear project description and rationale
- Including installation instructions for both Rust and npm/yarn
- Providing usage examples in Rust and TypeScript
- Adding a configuration section with code samples
- Including important security considerations
- Improving overall structure and readability

FabianLars commented:
sorry for the delay here, i'll try to get back to you as soon as i can

FabianLars commented on README.md:
```diff
@@ -1,8 +1,137 @@
+## Why This Plugin?
+
+Many OAuth providers (like Google and GitHub) don't allow custom URI schemes ("deep links") as
+redirect URLs for security reasons. This plugin provides a solution by:
```
```suggestion
redirect URLs. This plugin provides a solution by:
```
tbh feels more like most of them don't support it just because. I'd prefer not to say it's for security reasons so that i'm not asked to explain what those are 😅

FabianLars commented on README.md:
````diff
@@ -1,8 +1,137 @@
+```toml
+# Cargo.toml
+[dependencies]
+tauri-plugin-oauth = "0.0.0-alpha.0"
````
```suggestion
tauri-plugin-oauth = "2.0.0-rc.0"
```

FabianLars commented on README.md:
````diff
@@ -1,8 +1,137 @@
+For Tauri projects using npm or yarn:
+
+```bash
+npm install @fabianlars/tauri-plugin-oauth
+# or
+yarn add @fabianlars/tauri-plugin-oauth
````
```suggestion
npm install @fabianlars/tauri-plugin-oauth@2.0.0-rc.0
# or
yarn add @fabianlars/tauri-plugin-oauth@2.0.0-rc.0
```

FabianLars commented on README.md:
```diff
@@ -1,8 +1,137 @@
+#[cfg_attr(mobile, tauri::mobile_entry_point)]
+pub fn run() {
+    tauri::Builder::default()
+        .plugin(tauri_plugin_shell::init())
```
```suggestion
```

FabianLars commented on guest-js/index.ts:
````diff
@@ -0,0 +1,141 @@
+export interface OauthPlugin {
+  start: typeof start;
+  cancel: typeof cancel;
+  onUrl: typeof onUrl;
+  onInvalidUrl: typeof onInvalidUrl;
+}
+
+/**
+ * The main OAuth plugin object containing all plugin functions.
+ * @example
+ * ```typescript
+ * import { oauth } from '@fabianlars/tauri-plugin-oauth';
+ *
+ * async function initOAuth() {
+ *   const port = await oauth.start({ ports: [8000, 8001] });
+ *   console.log(`OAuth server started on port ${port}`);
+ *
+ *   oauth.onUrl((url) => {
+ *     console.log('Received OAuth URL:', url);
+ *     // Process the OAuth URL...
+ *   });
+ *
+ *   oauth.onInvalidUrl((error) => {
+ *     console.error('Received invalid OAuth URL:', error);
+ *     // Handle the error...
+ *   });
+ *
+ *   // Later, to stop the server:
+ *   // await oauth.cancel(port);
+ * }
+ *
+ * initOAuth();
+ * ```
+ */
+export const oauth: OauthPlugin = {
+  start,
+  cancel,
+  onUrl,
+  onInvalidUrl,
+};
````
hmm i think we should just export the methods directly, matching most of the official plugins.

cuervolu committed:
Update README.md

Co-authored-by: Fabian-Lars <github@fabianlars.de>

cuervolu committed:
Update tauri-plugin-oauth version in README.md

Co-authored-by: Fabian-Lars <github@fabianlars.de>

cuervolu committed:
Update install commands in README.md

Co-authored-by: Fabian-Lars <github@fabianlars.de>

cuervolu committed:
refactor(bindings): Refactor OAuth plugin exports

cuervolu committed:
chore(docs): Update OAuth plugin example in README.md

cuervolu commented:
Thank you for the suggestion. I've updated the code to export the methods directly. 

I've also updated the README and usage examples to reflect this change. Please let me know if you'd like to see any further modifications.

FabianLars commented:
Only the example needs to be updated for the new exports, then i think it's good to go. Thanks again :)

cuervolu committed:
refactor(example): Update OAuth plugin imports in example

cuervolu commented:
Oh, I forgot! But I’ve already fixed it, and at least on my PC everything works fine. Thanks for the reminder!

FabianLars committed:
Merge branch 'v2' into v2
